app/backends/workflow-orchestration.ts

```diff
@@ -21,6 +22,34 @@ const RETRY_DELAY = 1000;
 // Must match where the service wrapper is mounting artifacts
 export const PATH_TO_CONTAINER_ARTIFACTS = '/tmp/metadata';
 
+/**
+ * Calculate the granule page limit for the current query-cmr work item.
+ * @param workItem - current query-cmr work item
+ * @param tx - database transaction to query with
+ * @param logger - a Logger instance
+ * @returns a number used to limit the query-cmr task or undefined
+ */
+async function calculateQueryCmrLimit(
+  workItem: WorkItem, 
+  tx,
+  logger: Logger): Promise<number> {
+  if (workItem?.scrollID) { // only proceed if this is a query-cmr step
+    const nextStep = await getWorkflowStepByJobIdStepIndex(
+      tx,
+      workItem.jobID,
+      workItem.workflowStepIndex + 1,
+    );
+    if (!nextStep) {
+      logger.warn('Could not find next step for query CMR work item');
+      return;
+    }
+    const numItemsGenerated = await workItemCountForStep(tx, workItem.jobID, nextStep.stepIndex);
```

Comment:
Double-check this function works correctly for aggregated steps. For example if the next step is netcdf-to-zarr by default it will concatenate and so for a 5000 granule request we won't have any work items created for the next step yet, but we'll want to make sure the first query-cmr task gets 2000, the second gets 2000, and the third gets 1000. After that third request comes back, and we have all 5000 granules then we want to create a work item to send all 5000 to netcdf-to-zarr.

Reply:
OK -- will double check.